### ISSUE TITLE
Minimize data format?

### ISSUE THREAD

**Issue by greggman** [mdn/browser-compat-data#4527]
I'm sorry if this is a noob question and maybe a solution already exists but ....

It would be nice if the format could be more minimal. For example let's say there's a new API called `BrainConnection` and let's assume it has methods `start`, `stop` and `pause`. Let's assume only Firefox and Chrome support this API so far and further just Chrome doesn't yet support `pause`.

It would be nice to just be able to add the API and only those browsers that are known to support it with no requirement to check all browsers. In other words if only firefox and chrome have added support only they should be listed. Every other browser not listed is assumed to not support the feature. Devs or fans or others can add the other browsers when and if they support the feature removing the requirement that people trying to contribute be all knowing about all browsers.

Further, as mentioned above, let's assume Chrome doesn't support the `pause` API. Then it would be nice to just list chrome as not supporting the that feature. Any browser that's listed as supporting `BrainConnection` is assumed to support the entire API unless otherwise specified.

Effectively

```
{
  "api": {
    "BrainConnection": {
      "__compat": {
        "mdn_url": "https://developer.mozilla.org/docs/Web/API/BrainConnection",
        "support": {
          "chrome": {
            "version_added": "85"
          },
          "firefox": {
            "version_added": "82"
          }
      },
      "pause": {
        "__compat": {
          "mdn_url": "https://developer.mozilla.org/docs/Web/API/BrainConnection/pause",
          "support": {
            "chrome": {
              "version_added": false
            }
          }
        }
     }
  }
}
```

Or course then you have the issue that an added feature is assume to be available on all browsers but, you could add a flag to flip the expectations. Assume a new feature `simulate` is added. Then something like

```
      "simulate": {
        "__compat": {
          "mdn_url": "https://developer.mozilla.org/docs/Web/API/BrainConnection/stimulate",
          "feature_added": true,
          "support": {
            "firefox": {
              "version_added": "87"
            }
          }
        }
     }
```

Or alternatively you could derive that data by requiring the "version_added" data to be either all true/version or all false. In other words, if any "version_added" is "false" for a particular feature then only `"version_added": false` is allowed and would fail linting. If any "version_added" is **not** false then no "version_added" is allowed to be false.

In other words, you can easily tell if a feature is showing which browsers support the feature OR which browsers **do not** support the feature by checking the "version_added" values. If they are all false then that feature is showing which browsers don't support the feature. If they are all not false then that feature is showing which browsers **do** support the feature. If they are some combination of false and not false it's an error

The point of this suggestion is as it the format seems cumbersome to update. Let's say someone is very familiar with Firefox so they have some idea of what features it supports. It's unlikely they have any interest in finding out if Samsung's browser supports anything. It's not their expertise and expecting them to go dig up that info seems like a huge barrier to contributing to this database.

Changes like the ones suggested above seem like they'd make it much easier to contribute and therefore more likely to keep the data up to date.

**Comment by bershanskiy** [comment]
@greggman If you have a specific contribution in mind, but are stuck on figuring out the data schema, let's talk about that instead. Most likely, we do not need to change the schema to accommodate one more browser feature.

**TL;DR: A few words from a current outside contributor (not Mozilla or Google or another stakeholder employee) in support of current schema (against these changes). Using `null`s seems sufficient for cases.**

Current schema is simple to write and use; if data is missing a `"version_added"` value for a certain value, it is just `null` (or in some cases just omitted so evaluates to `undefined`). Easy to read for humans and consume in scripts.

> It would be nice if the format could be more minimal.

Keep in mind that shorter text is not necessarily easier to read than the longer one, especially if you make it shorter with shortcuts, like the one you propose. E.g., people prefer to write scripts in Python over Bash or Perl, even though you can write a complete program in the latter two.

I think what you actually want to say is "It would be nice if the _files would be shorter_." Long (1K+ lines) files are indeed poorly supported by GitHub online editor and might be hard to navigate, but you can already split features into sub-features and put them into separate files. The schema says the same, the data is just split into different files.

> `feature_added`

I'm still unclear on what `feature_added` would mean. Logically, browsers started out with a small set of features and every feature was added at some point. On a more pragmatic side, every feature goes through prototyping stages and developers concurrently draft W3C specifications and implement the features. All new and recent W3C drafts reach the "Technical Recommendation" status _only after_ there are two independent and complete implementations of the specification. So do you label absolutely everything with `feature_added`? After you inserted `feature_added`, when do you remove it?

> removing the requirement that people trying to contribute be all knowing about all browsers.

As far as I know, there is no formal requirement to include compatibility data for all browsers. People just try their best to make as complete PRs as possible. For example, you might have noticed Chrome and Firefox are very well covered (_way better_ than anywhere I know, including [caniuse](//caniuse.com)), while Samsung Internet is very underrepresented.

> Changes like the ones suggested above seem like they'd make it much easier to contribute and therefore more likely to keep the data up to date.

These changes might simplify adding a new feature to the data -- just insert one `"feature_added": true` and you are done, but they would make it harder to process data in scripts (you'd have to make a special case for it) and read the data (you'd have to constantly keep in mind when you saw `feature_added` the last time). In contrast, current schema is very explicit and you can just use `null`s to achieve the desired result. The proposed changes only complicate the schema and would only deter me from contributing in the first place.

**Comment by greggman** [comment]
I guess I didn't make myself clear.

I have no interest in the files being smaller. What I have as an interest in minimizing the work required to contribute and therefore making it easier to contribute.

First off my explanation for "feature_added" was clearly poor and I explained below that an implicit solution

The point is you either want to know

1. all browser that support a feature 
2. all browsers that DON'T support a feature.

In my case I wanted to document that of all the browsers that support AudioContext, Safari in particular does not support 2 features because of outstanding and 3 to 8 yr old bugs that they refuse to fix. To do this I had to add data for 9 other browsers that are not Safari. I had no interest in documenting those browsers nor do I have any expertise in those browser. I don't own them. I don't use them. I don't even have devices that run some of them. I don't know where on the internet or their official sites I'd look up the data for them. But, when I left them out the linter complained forcing me to add data I'm not qualified to add.

This will only get worse. At some point other browsers will get enough market share that they get added to the required list for the linter and people trying to contribute will have to add entries for 12 or 15 or 25 browsers when they actually only have knowledge of 1 browser. Effectively the current format is unscalable and when a new browser is added, especially according to linter, nearly every single file in the entire database needs to be updated in tons of places.

As I pointed out above, with a little thought these 2 types of data

1. all browser that support a feature 
2. all browsers that DON'T support a feature.

Are mutually exclusive and implicitly derivable. The browsers listed for a feature would be exclusively version_added=true/version in which case its (1) documenting browsers that support a feature, OR, version_added=false in which case its (2) documenting browsers that don't support a feature.

As for reading being harder if you believe that's actually a problem then some build step can generate a normalized version of the data from a concise version of the data. Make that part of the build process if you want so it's checked in here by some CI or whatever.

**Comment by ddbeck** [comment]
Hi @greggman, thank you for your thoughts on this. I'm definitely interested in how we can reduce contributor friction. You've got several ideas here, but I'm not sure I understand the problems you faced which inspired them. Before commenting on the schema proposals, I'd like to know more about the difficulty you've had with the schema as it is today.

For example, you talked about this:

> But, when I left them out the linter complained forcing me to add data I'm not qualified to add.

Would you be willing share an example of the structure you used and the linter errors you got?

I know that we require the full browser list to appear in every `__compat`, but I'm only aware of one rather limited case where we _require_ a non-`null` `"version_added"` value (that would be IE CSS entries). If the linter is complaining in other contexts, that suggests to me that there's a bug, poor documentation, or both (which we ought to fix, whether or not we change the schema).

**Comment by Elchi3** [comment]
We haven't heard back here. Closing.
